### DISCUSSION TITLE
getting "Future attached to a different" loop error

### DISCUSSION THREAD

**Discussion by tigrankh** [fastapi-users/fastapi-users#663] [Q&A]
Hi,

Thanks for all the work on fastapi-users.
I'm just starting with it, but the road began kind of bumpy :/

I'm running the Full Example posted in the docs and just added a uvicorn to run the app.
when sending a request to auth/jwt/login I get this error:

`RuntimeError: Task <Task pending name='Task-15' coro=<RequestResponseCycle.run_asgi() running at /Users/tigrankhazhakyan/dev/playground/fastapi-users/venv/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py:369> cb=[set.discard()]> got Future <Future pending cb=[_chain_future.<locals>._call_check_cancel() at /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/futures.py:360]> attached to a different loop`

The error is thrown when awaiting for get_by_email() method

the thing is that I don't explicitly create new loops, so the only existing loop I assume should be the main one.
I tried adding my own loop to motor class constructor, but it didn't change anything.

Can you please help?

Thanks,
Tigran

**Comment by frankie567** [comment]
Hi Tigran,

This kind of error can happen in some situations where several loops are running, like in unit tests for example. I'm surprised you get this using the full example. Would you mind sharing your code so we can check if there is something wrong? And also give us the command you use to run Uvicorn?

For the record, I've setup a working Repl.it using the full example: https://replit.com/@frankie567/fastapi-users-mongodb

**Reply by tigrankh** [reply]
Hi Frank,

Sorry for the delay in reply.
I found the problem. it appears when I run uvicorn from inside the python script, I can this error. assuming that probably uvicorn creates the second loop, but the futures were defined in the main look.

The error disappeared when I run uvicorn from the terminal.

Does that make sense to you?

Thanks,
Tigran

**Reply by frankie567** [reply]
Hmm... I made some tests and I found out it happens when you directly pass the `app` instance instead of the path module. This works:

```py
import uvicorn

if __name__ == "__main__":
    uvicorn.run("app:app", host="0.0.0.0", port=5000, log_level="info")
```

This provokes the "attached to a different loop" error:

```py
import uvicorn
from app import app

if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=5000, log_level="info")
```

My theory is the following:

* When we import `app` from the `app` module, it automatically loads the `AsyncIOMotorClient` instance, which is binded to the default asyncio loop.
* When Uvicorn starts, it [automatically creates a new loop](https://github.com/encode/uvicorn/blob/3a5c3b871379236f2867d7799d3480b94fcc5e60/uvicorn/loops/asyncio.py) and set it as the new default.
* Problem: Motor was already instantiated with another loop; hence the error.

When we let Uvicorn import the app by itself (with the `"app:app"` argument), it also creates a new loop and set it by default; but it imports Motor after, so they are binded to the same loop.

Another solution is to prevent Uvicorn to setup the loop by itself using the `loop` argument, but I don't know what could be the potential gotchas around this:

```py
import uvicorn
from app import app

if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=5000, log_level="info", loop="none")
```

My final advice:
* Prefer the command line invokation
* If you really need to invoke it programmatically, put it in another file and use the import style instead of importing the application yourself.

**Reply by tigrankh** [reply]
thanks for the answer. 
indeed I was starting uvicorn from a separate file, but by just telling "app" in the param list.

and thanks for suggestions, thats helpful

**Reply by scoutinhobh** [reply]
is this a bug? Should it matter how uvicorn is started?

**Comment by victorgerardodominguez** [comment]
Hello! i am having the same error but in this project i can run it in local with uvicron but when i run in  a container using gunicorn it shows up :(
When sending the request to auth, I get the error:

`RuntimeError: Task <Task pending coro=<RequestResponseCycle.run_asgi() running at /usr/local/lib/python3.7/site-packages/uvicorn/protocols/http/httptools_impl.py:376> cb=[set.discard()]> got Future <Future pending cb=[_chain_future.<locals>._call_check_cancel() at /usr/local/lib/python3.7/asyncio/futures.py:351]> attached to a different loop`

dockerfile: 

`FROM tiangolo/uvicorn-gunicorn-fastapi:python3.7
ENV PORT 8080
ENV APP_MODULE app.api:app
ENV LOG_LEVEL debug
ENV TIMEOUT 600


ADD . /app
WORKDIR /app
RUN pip install --no-cache-dir -r requirements.txt

EXPOSE 8080`


main.py:

`import os
import uvicorn
#from app import app

if __name__ == '__main__':
    #this will not work with multiprocessing
    uvicorn.run("app:app", host='0.0.0.0', port=8080, log_level="info")`


I also share the dockerfile and the main.py and finally comment that I was trying python 3.7 and 3.9 and I get the same thing.

Greetings and thanks

Victor
 
PS: I am connected to a database in mongodb atlas

**Comment by pasfrag** [comment]
@victorgerardodominguez did you find any solution on your problem? I have the same error on a server using nginx and guvicorn.

**Comment by rbrisita** [comment]
The issue could also be creating `asyncio.Queue()` outside of the current event loop. These queues should be created inside the main coroutine so they access the current running loop.

**Comment by tapan-nerospirit** [comment]
I am also facing same problems while storing data in the mongo @tigrankh @frankie567
